odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,599 @@
+        """
+        if self.domain.field == ComplexNumbers():
+            raise NotImplementedError('operator not Frechet-differentiable '
+                                      'on a complex space.')
+
+        if self.exponent == float('inf'):
+            raise NotImplementedError('operator not Frechet-differentiable '
+                                      'for exponent = inf.')
+
+        vf = self.domain.element(vf)
+        vf_pwnorm_fac = self(vf)
+        vf_pwnorm_fac **= (self.exponent - 1)
+
+        inner_vf = vf.copy()
+
+        for gi in inner_vf:
+            gi /= vf_pwnorm_fac * gi ** (self.exponent - 2)
+
+        return PointwiseInner(self.domain, inner_vf, weight=self.weights)
+
+
+class PointwiseInner(PointwiseTensorFieldOperator):
+
+    """Take the point-wise norm of a vector field.
+
+    This operator takes the (weighted) inner product
+
+        ``<F(x), G(x)> = sum_j ( w_j * F_j(x) * conj(G_j(x)) )
+
+    for a given vector field ``G``, where ``F`` is the vector field
+    acting as a variable to this operator.
+
+    This implies that the `Operator.domain` is a power space of a
+    discretized function space. For example, if ``X`` is a `DiscreteLp`
+    space, then ``ProductSpace(X, d)`` is a valid domain for any
+    positive integer ``d``.
+    """
+
+    def __init__(self, vfspace, vecfield, weight=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        vfspace : `ProductSpace`
+            Space of vector fields on which the operator acts.
+            It has to be a product space of identical spaces, i.e. a
+            power space.
+        vecfield : domain `element-like`
+            Vector field with which to calculate the point-wise inner
+            product of an input vector field
+        weight : `array-like` or `float`, optional
+            Weighting array or constant for the norm. If an array is
+            given, its length must be equal to ``domain.size``, and
+            all entries must be positive. A provided constant must be
+            positive.
+            By default, the weights are is taken from
+            ``domain.weighting``. Note that this excludes unusual
+            weightings with custom inner product, norm or dist.
+
+        Examples
+        --------
+        We make a tiny vector field space in 2D and create the
+        point-wise inner product operator with a fixed vector field.
+        The operator maps a vector field to a scalar function:
+
+        >>> import odl
+        >>> spc = odl.uniform_discr([-1, -1], [1, 1], (2, 2))
+        >>> vfspace = odl.ProductSpace(spc, 2)
+        >>> fixed_vf = np.array([[[0, 1], [-1, 1]],
+        ...                      [[1, -1], [1, -1]]])
+        >>> pw_inner = PointwiseInner(vfspace, fixed_vf)
+        >>> pw_inner.range == spc
+        True
+
+        Now we can calculate the inner product in each point:
+
+        >>> x = vfspace.element([[[1, 2], [0, 3]],
```

Comment:
I know, but it doesn't look any better with one row per line. Hard to make clear that we're creating a product space vector with 2 components, each 2x2.